orchestrator/test/conftest.py

```diff
@@ -125,3 +127,23 @@ def in_tmp_dir(tmp_path: Path) -> Path:
     os.chdir(tmp_path)
     yield tmp_path
     os.chdir(cwd)
+
+
+@pytest.fixture()
+def app_setup(request, device: Device, support_app: str, support_test_app: str, test_butler_service: str
+              ) -> Tuple[Application, TestApplication, ServiceApplication]:
+    uninstall_apk(support_app, device)
+    uninstall_apk(support_test_app, device)
+    uninstall_apk(test_butler_service, device)
+
+    support_application = Application.from_apk(support_app, device)
+    support_test_application = TestApplication.from_apk(support_test_app, device)
+    test_butler_service_application = ServiceApplication.from_apk(test_butler_service, device)
```

Comment:
Seems weird to do all three apps in one fixture -- every test only uses one, right?

You could make a generic 'package_installer' fixture which returns a function that takes the app and does the install and cleanup. https://docs.pytest.org/en/latest/fixture.html#factories-as-fixtures

Reply:
There's a test that uses the support app and the butler app (test_device_class.py::TestAndroidDevice::test_start_stop_app). I just figured it would be easier to have one place to call that would handle all of the apps, but I can split this into 3 separate fixtures.

Reply:
What i mean with the generic fixture is this:
    
    @pytest.fixture
    def install_app(device)
      apps = []
      def do_install(app_cls, package):
        uninstall_apk(package, device)
        app = app_cls.from_apk(package, device)
        apps.append(app)
        return app

      yield do_install
      for app in apps:
        app.uninstall()

Then in your test:

    def test_foo(install_app, support_test_app):
      install_app(TestApplication, support_test_app)
      ...

Reply:
That way you can do install_app as many times as you want, with whatever applications you need, and it still gets cleaned up at test end. No need for copy / pasting fixtures.

Reply:
You could also just change support_test_app and friends to be
    
    @pytest.fixture
    def support_test_app(install_app):
      return install_app(support_test_app_pkg)